src/backend/distributed/planner/recursive_planning.c

```diff
@@ -316,6 +351,189 @@ RecursivelyPlanCTEs(Query *query, RecursivePlanningContext *planningContext)
 }
 
 
+/*
+ * RecursivelyPlanSubqueryWalker recursively finds all the Query nodes and
+ * recursively plans if necessary.
+ */
+static bool
+RecursivelyPlanSubqueryWalker(Node *node, RecursivePlanningContext *context)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, Query))
+	{
+		Query *query = (Query *) node;
+		DeferredErrorMessage *error = NULL;
+
+		context->level += 1;
+		error = RecursivelyPlanSubqueriesAndCTEs(query, context);
+		if (error != NULL)
+		{
+			RaiseDeferredError(error, ERROR);
+		}
+		context->level -= 1;
+
+		if (ShouldRecursivelyPlanSubquery(query, context))
+		{
+			RecursivelyPlanSubquery(query, context);
+		}
+
+		return false;
+	}
+
+	return expression_tree_walker(node, RecursivelyPlanSubqueryWalker, context);
+}
+
+
+/*
+ * ShouldRecursivelyPlanSubquery decides whether the input subquery should be recursively
+ * planned or not.
+ *
+ * For the details, see the cases in the function.
+ */
+static bool
+ShouldRecursivelyPlanSubquery(Query *subquery, RecursivePlanningContext *context)
+{
+	if (FindNodeCheckInRangeTableList(subquery->rtable, IsLocalTableRTE))
+	{
+		/*
+		 * Postgres can always plan queries that don't require distributed planning.
+		 * Note that we need to check this first, otherwise the calls to the many other
+		 * Citus planner functions would error our due to local relations.
+		 *
+		 * TODO: We could only successfully create distributed plans with local tables
+		 * when the local tables are on the leaf queries and the upper level queries
+		 * do not contain any other local tables.
+		 */
+	}
+	else if (DeferErrorIfCannotPushdownSubquery(subquery, false) == NULL)
+	{
+		/*
+		 * Citus can pushdown this subquery, no need to recursively
+		 * plan which is much expensive than pushdown.
+		 */
+		return false;
+	}
+	else if (TaskExecutorType == MULTI_EXECUTOR_TASK_TRACKER &&
+			 SingleRelationRepartitionSubquery(subquery))
+	{
+		return false;
+	}
+
+	/*
+	 * Even if we could recursively plan the subquery, we should ensure
+	 * that the subquery doesn't contain any references to the outer
+	 * queries.
+	 */
+	if (ContainsReferencesToOuterQuery(subquery))
+	{
+		if (log_min_messages <= DEBUG2 || client_min_messages <= DEBUG2)
+		{
+			elog(DEBUG2, "skipping recursive planning for the subquery since it "
+						 "contains references to outer queries");
```

Comment:
I wonder whether it's worth showing the query here, because it's hard to know what this is about in a complex query. (consider this optional)

Reply:
I already tried that but deparsing queries with outer references fails (I guess due to `pull_var_clause` at some point while deparsing)

Reply:
ah, makes sense